src/SourceGenerators/Uno.UI.SourceGenerators/Helpers/HashBuilder.cs

```diff
@@ -28,25 +33,32 @@ public static string BuildIDFromSymbol(ITypeSymbol typeSymbol)
 		/// <param name="input">The string to hash</param>
 		/// <param name="algoritm">The algorithm to use</param>
 		/// <returns>A hex-formatted string of the hash</returns>
-		public static string Build(string input, HashAlgorithm algoritm = null)
+		public static string Build(string input)
 		{
-			using (algoritm ??= SHA256.Create())
+			using (var algoritm = SHA256.Create())
 			{
 				// Convert the input string to a byte array and compute the hash.
 				var data = algoritm.ComputeHash(Encoding.UTF8.GetBytes(input));
 
-				// Create a new Stringbuilder to collect the bytes
-				// and create a string.
-				var builder = new StringBuilder();
-
-				// Use the first 16 bytes of the hash
-				for (int i = 0; i < Math.Min(data.Length, 16); i++)
+				var min = Math.Min(data.Length, 16);
+				var result = new string((char)0, min * 2);
+				// efficient implementation to convert a byte array to hex string.
+				// See https://github.com/patridge/PerformanceStubs/blob/574c79441c18220f7841bdd1c25db0f6d7752876/README.markdown#converting-a-byte-array-to-a-hexadecimal-string
+				unsafe
 				{
-					builder.Append(data[i].ToString("x2", CultureInfo.InvariantCulture));
+					var lookupP = _lookup32UnsafeP;
```

Comment:
As an alternative and safer option, you could also check out what I'm doing in ComputeSharp ([here](https://github.com/Sergio0694/ComputeSharp/blob/ccd0a5da74fd45fd201938cf48caa76e26806065/src/ComputeSharp.D2D1.SourceGenerators/ID2D1ShaderGenerator.CreateLoadBytecodeMethod.Syntax.cs#L192)). You could just reuse that lookup table (removing the leading `0x` if you don't need them) and just index from there, without the need for pointers nor pinning at all. It'd still avoid allocating strings for each value in this loop just the same.

Reply:
Nice! I benchmarked and the lookup table approach has the same performance. I used it in the last commit.